File: desktop/packages/core/src/data/subject-repository.ts
import type { Database } from '../db/connection.js';
import { localUserId, newId } from '../ids.js';
import { fromEpochSeconds, toEpochSeconds } from '../time.js';
import type { SubjectRepository, SubjectRow } from './ports.js';

interface RawSubject {
  id: string;
  name: string;
  color: string | null;
  source: string;
  source_name: string | null;
  archived: number;
  created_at: number;
}

function toRow(r: RawSubject): SubjectRow {
  return {
    id: r.id,
    name: r.name,
    color: r.color,
    source: r.source,
    sourceName: r.source_name,
    // Stored as INTEGER 0/1 with a CHECK; exposed as a boolean so no caller
    // has to know that.
    archived: r.archived === 1,
    createdAt: fromEpochSeconds(r.created_at),
  };
}

const SELECT_COLUMNS = `id, name, color, source, source_name, archived, created_at`;

export class SqliteSubjectRepository implements SubjectRepository {
  constructor(
    private readonly db: Database,
    private readonly now: () => Date = () => new Date(),
  ) {}

  async create(a: {
    name: string;
    color?: string | null;
    source?: string;
    sourceName?: string | null;
  }): Promise<string> {
    const id = newId();
    const ts = toEpochSeconds(this.now());
    this.db
      .prepare(
        `INSERT INTO subjects
           (id, user_id, name, color, source, source_name, created_at, updated_at)
         VALUES (?, ?, ?, ?, ?, ?, ?, ?)`,
      )
      .run(
        id,
        localUserId,
        a.name,
        a.color ?? null,
        a.source ?? 'self',
        a.sourceName ?? null,
        ts,
        ts,
      );
    return id;
  }

  async update(
    id: string,
    a: { name: string; color: string | null; source: string; sourceName: string | null },
  ): Promise<void> {
    this.db
      .prepare(
        `UPDATE subjects
            SET name = ?, color = ?, source = ?, source_name = ?, updated_at = ?
          WHERE id = ?`,
      )
      .run(a.name, a.color, a.source, a.sourceName, toEpochSeconds(this.now()), id);
    return;
  }

  async setArchived(id: string, archived: boolean): Promise<void> {
    // node:sqlite refuses to bind a JS boolean (probed 2026-07-27), and schema
    // v1 stores this as INTEGER 0/1 anyway.
    this.db
      .prepare(`UPDATE subjects SET archived = ?, updated_at = ? WHERE id = ?`)
      .run(archived ? 1 : 0, toEpochSeconds(this.now()), id);
    return;
  }

  async remove(id: string): Promise<void> {
    const ts = toEpochSeconds(this.now());
    this.db
      .prepare(`UPDATE subjects SET deleted_at = ?, updated_at = ? WHERE id = ?`)
      .run(ts, ts, id);
    return;
  }

  async list(opts: { archived?: boolean } = {}): Promise<SubjectRow[]> {
    const rows = this.db
      .prepare(
        `SELECT ${SELECT_COLUMNS}
           FROM subjects
          WHERE deleted_at IS NULL AND archived = ?
          ORDER BY created_at DESC, id DESC`,
      )
      .all<RawSubject>(opts.archived === true ? 1 : 0);
    return rows.map(toRow);
  }
}

File: app-flutter/lib/features/subjects/data/subject_repository.dart
import 'package:drift/drift.dart';

import '../../../core/db/database.dart';
import '../../../core/ids.dart';

class SubjectRepository {
  SubjectRepository(this._db);

  final AppDatabase _db;

  Future<String> createSubject(
    String name, {
    String? color,
    String source = 'self',
    String? sourceName,
  }) async {
    final id = newId();
    await _db.into(_db.subjects).insert(SubjectsCompanion.insert(
          id: id,
          userId: localUserId,
          name: name,
          color: Value(color),
          source: Value(source),
          sourceName: Value(sourceName),
        ));
    return id;
  }

  /// Writes all editable fields; null clears colour / source name.
  Future<void> updateSubject(
    String id, {
    required String name,
    String? color,
    required String source,
    String? sourceName,
  }) async {
    await (_db.update(_db.subjects)..where((s) => s.id.equals(id))).write(
      SubjectsCompanion(
        name: Value(name),
        color: Value(color),
        source: Value(source),
        sourceName: Value(sourceName),
        updatedAt: Value(DateTime.now().toUtc()),
      ),
    );
  }

  Future<void> setArchived(String id, {required bool archived}) async {
    await (_db.update(_db.subjects)..where((s) => s.id.equals(id))).write(
      SubjectsCompanion(
        archived: Value(archived),
        updatedAt: Value(DateTime.now().toUtc()),
      ),
    );
  }

  /// Soft delete (data-model.md §2) — history joins keep the name.
  Future<void> deleteSubject(String id) async {
    final now = DateTime.now().toUtc();
    await (_db.update(_db.subjects)..where((s) => s.id.equals(id))).write(
      SubjectsCompanion(deletedAt: Value(now), updatedAt: Value(now)),
    );
  }

  Stream<List<Subject>> watchSubjects({bool archived = false}) {
    return (_db.select(_db.subjects)
          ..where((s) => s.deletedAt.isNull() & s.archived.equals(archived))
          ..orderBy([(s) => OrderingTerm.desc(s.createdAt)]))
        .watch();
  }
}
